File: frontend/src/api/metas.js
import { cliente } from './cliente.js'

// Las metas de ahorro y sus aportaciones.
//
// No se usa el crudDe de catalogos.js: el listado devuelve un objeto
// { metas, resumen } en vez de un arreglo, y ademas hay dos rutas mas para el
// sub-recurso de aportaciones. Forzarlo dentro del molde comun costaria mas
// explicaciones de las que ahorra.

export function listar(params, opciones) {
  return cliente.get('/metas', { params, ...opciones }).then((r) => r.data.datos)
}

// Devuelve la meta con el detalle de cada aportacion, que es lo que explica de
// donde sale el total.
export function obtener(id, opciones) {
  return cliente.get(`/metas/${id}`, opciones).then((r) => r.data.datos)
}

export function crear(datos) {
  return cliente.post('/metas', datos).then((r) => r.data.datos)
}

export function actualizar(id, datos) {
  return cliente.put(`/metas/${id}`, datos).then((r) => r.data.datos)
}

export function eliminar(id) {
  return cliente.delete(`/metas/${id}`).then(() => true)
}

// Las aportaciones cuelgan de su meta: no existen por su cuenta y por eso su
// ruta lleva dentro el identificador de la meta.
export function aportar(metaID, datos) {
  return cliente.post(`/metas/${metaID}/aportaciones`, datos).then((r) => r.data.datos)
}

export function quitarAportacion(metaID, aportacionID) {
  return cliente.delete(`/metas/${metaID}/aportaciones/${aportacionID}`).then(() => true)
}

File: backend/internal/handlers/metas.go
package handlers

import (
	"context"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/v2/bson"

	"github.com/FNTR3455234/FinTrack/backend/internal/modelos"
	"github.com/FNTR3455234/FinTrack/backend/internal/respuestas"
)

// ServicioMetas es lo que los handlers necesitan del servicio de metas.
type ServicioMetas interface {
	Crear(ctx context.Context, usuarioID bson.ObjectID, p modelos.PeticionMeta) (*modelos.Meta, error)
	Listar(ctx context.Context, usuarioID bson.ObjectID, f modelos.FiltroMetas) ([]modelos.ProgresoMeta, modelos.ResumenMetas, error)
	PorID(ctx context.Context, usuarioID, id bson.ObjectID) (*modelos.MetaConAportaciones, error)
	Actualizar(ctx context.Context, usuarioID, id bson.ObjectID, p modelos.PeticionMeta) (*modelos.Meta, error)
	Eliminar(ctx context.Context, usuarioID, id bson.ObjectID) error
	Aportar(ctx context.Context, usuarioID, metaID bson.ObjectID, p modelos.PeticionAportacion) (*modelos.Aportacion, error)
	QuitarAportacion(ctx context.Context, usuarioID, metaID, aportacionID bson.ObjectID) error
}

// Metas agrupa los handlers de /metas.
type Metas struct {
	servicio ServicioMetas
}

// NuevoMetas construye el grupo de handlers.
func NuevoMetas(servicio ServicioMetas) *Metas {
	return &Metas{servicio: servicio}
}

// Listar atiende GET /metas.
//
//	@Summary		Listar metas de ahorro
//	@Description	Cada meta con lo que se lleva ahorrado, lo que falta, el porcentaje, el estado y a que ritmo mensual hay que ahorrar para llegar a tiempo. Cruza metas con aportaciones (consulta relacional 3).
//	@Tags			metas
//	@Produce		json
//	@Security		BearerAuth
//	@Param			incluir_archivadas	query	bool	false	"Incluye tambien las archivadas"
//	@Success		200	{object}	respuestas.Sobre{datos=modelos.ListadoMetas}
//	@Failure		401	{object}	respuestas.SobreError
//	@Router			/metas [get]
func (h *Metas) Listar(c *gin.Context) {
	usuarioID, ok := usuarioAutenticado(c)
	if !ok {
		return
	}

	filtro := modelos.FiltroMetas{IncluirArchivadas: booleano(c, "incluir_archivadas")}

	metas, resumen, err := h.servicio.Listar(c.Request.Context(), usuarioID, filtro)
	if err != nil {
		respuestas.Fallo(c, err)
		return
	}
	respuestas.OK(c, modelos.ListadoMetas{Metas: metas, Resumen: resumen})
}

// Obtener atiende GET /metas/:id.
//
//	@Summary		Obtener una meta con sus aportaciones
//	@Description	El progreso mas el detalle de cada aportacion, que es lo que explica de donde sale la cifra. Responde 404 si la meta es de otro usuario.
//	@Tags			metas
//	@Produce		json
//	@Security		BearerAuth
//	@Param			id	path	string	true	"Identificador de 24 caracteres"
//	@Success		200	{object}	respuestas.Sobre{datos=modelos.MetaConAportaciones}
//	@Failure		400	{object}	respuestas.SobreError
//	@Failure		401	{object}	respuestas.SobreError
//	@Failure		404	{object}	respuestas.SobreError
//	@Router			/metas/{id} [get]
func (h *Metas) Obtener(c *gin.Context) {
	usuarioID, ok := usuarioAutenticado(c)
	if !ok {
		return
	}
	id, ok := idDeLaRuta(c)
	if !ok {
		return
	}

	meta, err := h.servicio.PorID(c.Request.Context(), usuarioID, id)
	if err != nil {
		respuestas.Fallo(c, err)
		return
	}
	respuestas.OK(c, meta)
}
